resources/builder.inc.sh

```diff
@@ -169,35 +169,73 @@ function builder_heading() {
 
 
 builder_echo() {
-  local color=white message= mark=
-  if [[ $# -gt 1 && $1 =~ ^(white|grey|green|success|blue|heading|yellow|warning|red|error|purple|brightwhite|teal|debug|setmark)$ ]]; then
-    color="$1"
-    shift
+  local color=white message= mark= block= action= do_output=true is_running_on_teamcity=false
+  if [[ -n "${TEAMCITY_GIT_PATH:-}" ]]; then
+    is_running_on_teamcity=true
+  fi
+
+  if [[ $# -gt 1 ]]; then
+    if [[ $1 =~ ^(white|grey|green|success|blue|heading|yellow|warning|red|error|purple|brightwhite|teal|debug|setmark)$ ]]; then
+      color="$1"
+      shift
+    elif [[ $1 == "start" ]]; then
+      # builder_echo start block message
+      action="$1"
+      block="$2"
+      shift 2
+      color="heading"
+      if ${is_running_on_teamcity}; then
+        do_output=true
+      else
+        do_output=${_builder_debug_internal:-false}
+      fi
+    elif [[ $1 == "end" ]]; then
+      # builder_echo end block status message
+      action="$1"
+      block="$2"
+      color="$3"
+      shift 3
+      if [[ "${color}" == "error" ]] || ${is_running_on_teamcity}; then
```

Comment:
I am not sure that this is correct. I would keep the _builder_execute_child calls to this unchanged, and then remove this logic. We do not normally emit for calling child actions because the child script would do its own emits anyway, which is why they are debug-conditioned.

That should make this PR significantly simpler:
* no need to track `do_output`
* no change to `_builder_execute_child`

Reply:
and here is an enhanced version of the original PR that also outputs the script name ([link](https://build.palaso.org/buildConfiguration/KeymanLinux_TestPullRequests/543724?buildTab=log&focusLine=2159&logView=linear&linesState=160.215.218.222)):

![image](https://github.com/user-attachments/assets/cad21098-019f-4f96-b9e9-5cf06a611b72)